### ISSUE TITLE
<chrono>: clock_cast throws an exception under windows7 sp1

### ISSUE THREAD

**Issue by itmanfei** [microsoft/STL#2446]
code:
`try {
  auto timestamp = std::chrono::clock_cast<std::chrono::system_clock>(std::filesystem::last_write_time(entity));
} catch (const std::exception& e) {
  // Caught exception
}`

debug:
`__std_tzdb_get_time_zones` failed because the error code is `_Win_error`

**Comment by fsb4000** [comment]
>While the STL generally provides all features on all supported versions of Windows, leap seconds and time zones (which change over time) require OS support that was added to Windows 10. Specifically, updating the leap second database requires Windows 10 version 1809 or later, and time zones require icu.dll which is provided by Windows 10 version 1903/19H1 or later. This applies to both client and server OSes; note that Windows Server 2019 is based on Windows 10 version 1809.


See:

1) `clock_cast` calls `clock_time_conversion` https://github.com/microsoft/STL/blob/9265c51c86b28ce67e82b3bdd9d270798cfa1ca3/stl/inc/chrono#L3584-L3585

2)`clock_time_conversion` calls `utc_clock::to_sys` 
https://github.com/microsoft/STL/blob/9265c51c86b28ce67e82b3bdd9d270798cfa1ca3/stl/inc/chrono#L3428-L3434

3) `utc_clock::to_sys`  calls `get_leap_second_info`
https://github.com/microsoft/STL/blob/9265c51c86b28ce67e82b3bdd9d270798cfa1ca3/stl/inc/chrono#L3208-L3211

BOOM.

I'm not a chrono expert. Perhaps something in this sequence of calls is wrong.  But in my opinioin it's unfortunate but i's correct.

Now is the holiday period at Microsoft, until January 3 (or 10 for some people). So the response from `chrono` maintainer may be delayed...

**Comment by itmanfei** [comment]
> > While the STL generally provides all features on all supported versions of Windows, leap seconds and time zones (which change over time) require OS support that was added to Windows 10. Specifically, updating the leap second database requires Windows 10 version 1809 or later, and time zones require icu.dll which is provided by Windows 10 version 1903/19H1 or later. This applies to both client and server OSes; note that Windows Server 2019 is based on Windows 10 version 1809.
> 
> See:
> 
> 1. `clock_cast` calls `clock_time_conversion` https://github.com/microsoft/STL/blob/9265c51c86b28ce67e82b3bdd9d270798cfa1ca3/stl/inc/chrono#L3584-L3585
> 
> 2) calls `clock_time_conversion``utc_clock::to_sys`
> 
> https://github.com/microsoft/STL/blob/9265c51c86b28ce67e82b3bdd9d270798cfa1ca3/stl/inc/chrono#L3428-L3434
> 
> 3. `utc_clock::to_sys`  calls `get_leap_second_info`
>    https://github.com/microsoft/STL/blob/9265c51c86b28ce67e82b3bdd9d270798cfa1ca3/stl/inc/chrono#L3208-L3211
> 
> BOOM.
> 
> I'm not a chrono expert. Perhaps something in this sequence of calls is wrong. But in my opinioin it's unfortunate but i's correct.
> 
> Now is the holiday period at Microsoft, until January 3 (or 10 for some people). So the response from maintainer may be delayed...`chrono`

Thanks for your answer.
You are right, I also saw the implementation of tzdb.cpp today. Icu.dll needs to be loaded, but there is no icu.dll and its dependent dll under windows7.

**Comment by MattStephanson** [comment]
Yes, the clock conversions that need leap second info take it from the `tzdb_list`, which will throw during construction if icu.dll isn't found. Probably the easiest approach would be to create a `tzdb` without any time zones, but that could break existing code that expects an exception in that case. (I find [[time.zone.db]](http://eel.is/c++draft/time.zone.db) unclear or silent about whether that would be a "valid" `tzdb` or if `current_zone()` may return `nullptr` or `throw`.) We could also have an internal function that just generates leap second data if a `tzdb` isn't available, but it would have to either be recreated every time it's needed or cached in a singleton in the manner of `tzdb_list`.

**Comment by cpplearner** [comment]
See also #1911

**Comment by doomlaur** [comment]
For now we built a wrapper function to benefit from high-precision file write times from std::filesystem, while still supporting older versions of Windows:
```cpp
// Define these somewhere
using file_duration = std::chrono::duration<long long, std::filesystem::file_time_type::period>;
using file_time_point = std::chrono::time_point<std::chrono::system_clock, file_duration>;

// Write a wrapper function that avoids clock_casts
// Note: leap seconds are NOT considered, as they require a time zone database!
std::error_code ec;
auto file_time = std::filesystem::last_write_time(path, ec);
auto last_write_time = std::chrono::time_point_cast<file_duration>(file_time - std::filesystem::file_time_type::clock::now() + std::chrono::system_clock::now());
```

**Comment by YexuanXiao** [comment]
If the first VS18 version will no longer support systems prior to Windows 10, then this issue can be closed. @StephanTLavavej

**Comment by trueqbit** [comment]
Adding another workaround to @doomlaur's.

I also think it's very aggressive of the STL team to silently drop older operating systems, especially since Server 2019 1809 is not that old yet. But it seems that an internal fallback is not easy to implement as the conversion from and to file time is done via UTC.

There seems to be a surprisingly easy workaround: Because `FILETIME` only became leap-second-aware as of Server 2019 and Windows 10 October update (Build 1903), a fallback to good old epoch difference computation can be done conditionally with according helpers.
Simple epoch difference is also what the `winrt::clock` is (still) doing.

There's a [class namespace](https://gitlab.com/klaustriendl/kj-cpp/-/blob/d081cf2e720b7757e689f478cf3a82a87d93532d/kj++/fileclock_fixup.h) as the workaround that I am using; sharing it here inline for reference:

```c++
// Copyright klaus triendl and piece of code LLC <klaus@triendl.eu>
#pragma once
#include <chrono>
#include <type_traits>

namespace kj {

/** @short Perform file clock time point conversions on systems where no time
   zone database is available.

   The `fileclock_fixup` class namespace provides the static utility methods
   `to_sys()` and `from_sys()` that use `std::chrono::clock_cast<>()` if the
   time zone database could be loaded, or fall back to the fixed conversion
   otherwise.

   The availability is subject to the system and the C++ standard library
   implementation, whose details are a bit convoluted.

   The Microsoft STL team decided to use leap second calculations
   unconditionally for C++20 chrono clock conversions starting with VC 17.13.
   For this to work, during instantiation of the `std::chrono::tzdb_list`
   singleton, they load icu.dll, which is only an integral part since Windows
   10, version 1903 (Anniversary Update) and later. This is also the OS version
   that accomodates for leap seconds in their Windows file time based APIs. The
   STL team also decided not use use a fallback on earlier systems. In short,
   C++20 chrono requires Windows 10, version 1903 or later.

   The good news is that on earlier systems, the file clock is not aware of leap
   seconds and uses a fixed offset from the system clock epoch. The offset is
   -11644473600 seconds, which is the difference between the Windows FILETIME
   epoch (Jan 1, 1601) and the Unix epoch (Jan 1, 1970).

   By the way, these functions could be used to write a custom file clock class,
   similar to the `winrt::file_clock`, see `windows_file_clock` below.
 */
struct fileclock_fixup {
  /** @short Convert a file clock time point to a system clock time point.
   */
  template <typename Duration>
  static std::chrono::sys_time<
      std::common_type_t<Duration, std::chrono::seconds>>
  to_sys(const std::chrono::file_time<Duration> &fileTime) noexcept {
    using namespace std::chrono;

    if (_tzdb) {
      return clock_cast<system_clock>(fileTime);
    }
    // fallback
    else {
      return _offset_sys_epoch + fileTime.time_since_epoch();
    }
  }

  /** @short Convert a system clock time point to a file clock time point.
   */
  template <typename Duration>
  static std::chrono::file_time<
      std::common_type_t<Duration, std::chrono::seconds>>
  from_sys(const std::chrono::sys_time<Duration> &ticks) noexcept {
    using namespace std::chrono;

    if (_tzdb) {
      return clock_cast<file_clock>(ticks);
    }
    // fallback
    else {
      using result_type = file_time<std::common_type_t<Duration, seconds>>;

      return result_type{ticks - _offset_sys_epoch};
    }
  }

  // system_clock epoch is 00:00:00, Jan 1 1970.
  // This is 11644473600 seconds after Windows FILETIME epoch of
  // 00:00:00, Jan 1 1601.
  static constexpr std::chrono::sys_seconds _offset_sys_epoch{
      std::chrono::seconds{-11644473600}};

  static std::chrono::tzdb_list *_fetch_tzdb() noexcept {
    try {
      return &std::chrono::get_tzdb_list();
    } catch (const std::runtime_error &) {
      return nullptr;
    };
  };

  static inline const std::chrono::tzdb_list *const _tzdb = _fetch_tzdb();
};

}
```

```c++
int main() {
  using namespace std::chrono;
    
  system_clock::time_point sysTimePoint = kj::fileclock_fixup::to_sys(file_clock::now());
}
```
